### ISSUE TITLE
Multi model validation.

### ISSUE THREAD

**Issue by sl33kr** [davidkpiano/react-redux-form#562]
Hi,
We have a validation requirement of user has entered Phone Number OR Email.

What is the idiomatic way to implement this in react-redux-form? The [validation examples](https://davidkpiano.github.io/react-redux-form/docs/guides/validation.html) really only cater for a single model value.

Should I handle the rff/changed action in a validation reducer and then dispatch the outcome of the validation?

**Comment by davidkpiano** [comment]
Take a look at the example in [the Form documentation:](https://davidkpiano.github.io/react-redux-form/docs/api/Form.html)

```jsx
import { Form, Control } from 'react-redux-form';
import { isEmail, isNull } from 'validator';

const required = isNull;

const passwordsMatch = ({ password, confirmPassword }) => {
  return password === confirmPassword;
};

// in render() return block:
<Form model="user"
  validators={{
    '': { passwordsMatch },
    email: { required, isEmail },
    password: { required },
    confirmPassword: { required }
  }}
  onSubmit={...}>
  <Control
    type="email"
    model=".email"
  />

  <Control
    type="password"
    model=".password"
  />

  <Control
    type="password"
    model=".confirmPassword"
  />

  <button>Submit!</button>
</Form>
```

This is an example of a **form-level validator** that takes in the entire form value and validates that both the `password` and `confirmPassword` models match.

So I'd imagine you'd have a `phoneNumberOrEmail` form-level validator that looks something like this:

```
const phoneNumberOrEmail = ({ phone, email }) => phone || email;

<Form model="user" validators={{
  '': { phoneNumberOrEmail }
}}>
// ...
```

I'll add a form-level validation example to the docs. Thanks for pointing it out!

**Comment by sl33kr** [comment]
Ah sorry! I do remember seeing this now.

**Comment by sl33kr** [comment]
Ok I got it working, is there a way to associate this validation with the field models though?

I can display then under the controls just by adding
```
<Errors model="appState.enquiry" messages={{ hasPhoneOrEmail: 'Please enter either a phone number or email.' }} />
```

but would prefer to only display it after either of the fields had been touched.

When I have some more time I'll have a dig through the source.

**Comment by davidkpiano** [comment]
> but would prefer to only display it after either of the fields had been touched.

Easy enough! `<Errors model="..." messages={{...}} show="touched" />` ;-)

> is there a way to associate this validation with the field models though?

Not quite yet - I'm going to add support for functions-as-validators in `<Form>` soon (they're available on `<Control>` right now). That way, you can just do this:

```jsx
<Form model=".." validators={(values) => {
  const phoneEmailValid = hasPhoneOrEmail(values);

  return {
    phone: { phoneEmailValid }, // associate with just the phone field?
    // ... other validity
  };
}}/>
```

for any super-custom behavior.

**Comment by sl33kr** [comment]
Hi David,
I realized I closed this but thought I would just add this in case someone came across this ticket.

Using `show="touched"` in the Errors component when it is display Form errors will cause the validation to display as soon as ANY field in the form is touched.

So in our case we had it display under phone/email half way down the form. As soon as the user tabbed from the first field, the phone/email error would display.
